src/ui/public/utils/subscribe.test.ts

```diff
@@ -0,0 +1,143 @@
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+const mockFatalError = jest.fn();
+jest.mock('ui/notify/fatal_error', () => ({
+  fatalError: mockFatalError,
+}));
+
+import * as Rx from 'rxjs';
+import { subscribe } from './subscribe';
+
+let $rootScope: Scope;
+
+class Scope {
+  public $$phase?: string;
+  public $root = $rootScope;
+  public $apply = jest.fn((fn: () => void) => fn());
+}
+
+$rootScope = new Scope();
+
+afterEach(() => {
+  jest.clearAllMocks();
+});
+
+it('subscribes to the passed observable, returns subscription', () => {
+  const $scope = new Scope();
+
+  const unsubSpy = jest.fn();
+  const subSpy = jest.fn(() => unsubSpy);
+  const observable = new Rx.Observable(subSpy);
+
+  const subscription = subscribe($scope as any, observable);
+  expect(subSpy).toHaveBeenCalledTimes(1);
+  expect(unsubSpy).not.toHaveBeenCalled();
+
+  subscription.unsubscribe();
+
+  expect(subSpy).toHaveBeenCalledTimes(1);
+  expect(unsubSpy).toHaveBeenCalledTimes(1);
+});
+
+it('calls observer.next() if already in a digest cycle, wraps in $scope.$apply if not', () => {
+  const subject = new Rx.Subject();
+  const nextSpy = jest.fn();
+  const $scope = new Scope();
+
+  subscribe($scope as any, subject, { next: nextSpy });
+
+  subject.next();
+  expect($scope.$apply).toHaveBeenCalledTimes(1);
+  expect(nextSpy).toHaveBeenCalledTimes(1);
+
+  jest.clearAllMocks();
+
+  $rootScope.$$phase = '$digest';
+  subject.next();
+  expect($scope.$apply).not.toHaveBeenCalled();
+  expect(nextSpy).toHaveBeenCalledTimes(1);
+});
+
+it('reports fatalError if observer.next() throws', () => {
+  const $scope = new Scope();
+  subscribe($scope as any, Rx.of(undefined), {
+    next() {
+      throw new Error('foo bar');
+    },
+  });
+
+  expect(mockFatalError.mock.calls).toMatchInlineSnapshot(`
+Array [
+  Array [
+    [Error: foo bar],
+  ],
+]
+`);
+});
+
+it('reports fatal error if observer.error is not defined and observable errors', () => {
+  const $scope = new Scope();
+  const error = new Error('foo');
+  error.stack = `${error.message}\n---stack trace ---`;
+  subscribe($scope as any, Rx.throwError(error));
+
+  expect(mockFatalError.mock.calls).toMatchInlineSnapshot(`
+Array [
+  Array [
+    [Error: Uncaught error in $scope.$subscribe: foo
+---stack trace ---],
+  ],
+]
+`);
+});
+
+it('reports fatal error if observer.error throws', () => {
```

Comment:
__nit:__ maybe one more test for `does not report fatal error if observer.error handles the error`?